table/tables/partition.go

```diff
@@ -180,29 +223,48 @@ func partitionRecordKey(pid int64, handle int64) kv.Key {
 func (t *partitionedTable) locatePartition(ctx sessionctx.Context, pi *model.PartitionInfo, r []types.Datum) (int64, error) {
 	var err error
 	var isNull bool
-	partitionExprs := t.partitionExpr.UpperBounds
-	idx := sort.Search(len(partitionExprs), func(i int) bool {
+	idx := 0
+	switch t.meta.Partition.Type {
+	case model.PartitionTypeRange:
```

Comment:
Extract two functions, like `locateRangePartition` ... If you want to reuse the range partition logic for hash partition
At least we can touch less code if later we rewrite the `locateHashPartition`